Automattic/wp-calypso
# Checkout: Simplify contact validation and add tests

#### Changes proposed in this Pull Request

This PR converts all the code that handles checkout contact validation to TypeScript, deduplicates the different ways that validation can be called, removes unnecessary dependencies like lodash and `wpcom.undocumented`, and simplifies a lot of the validation functions to make them more readable.

To guard against regressions, this adds integration tests for the four types of validation in checkout:

- "Tax" contact information validation.
- Domain contact validation (including CCTLD-specific fields).
- G Suite contact information validation.
- Logged-out signup email validation.

#### Testing instructions

This includes integration tests, but you can try the validation manually.

- Enter checkout with just a plan.
- Enter invalid postal code/country details and try to complete the form.
- Verify that the validation fails and that you get an error message displayed on the invalid field.
- Enter valid postal code/country details and try to complete the form.
- Verify that the validation succeeds.
- Repeat the above steps with a domain in the cart.
